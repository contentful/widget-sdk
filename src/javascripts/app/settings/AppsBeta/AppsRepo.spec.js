import createAppsRepo from './AppsRepo.es6';

const NETLIFY_APP_ID = 'netlify';
const NETLIFY_EXTENSION_DEFINITION_ID = '3VJXxF6XcYPl4akixQuJlc';

describe('AppsRepo', () => {
  describe('getApps', () => {
    // These tests are using Netlify App and Definition IDs
    // but they shouldn't break if we add more (we use only
    // `arrayContaining`, `objectContaining`, not equality).
    const netlifyDefinition = {
      sys: { type: 'ExtensionDefinition', id: NETLIFY_EXTENSION_DEFINITION_ID },
      name: 'Netlify',
      locations: ['app', 'entry-sidebar'],
      src: 'http://localhost:1234'
    };

    const netlifyExtension = {
      sys: { type: 'Extension', id: 'netlify-extension-id' },
      extensionDefinition: {
        sys: {
          type: 'Link',
          linkType: 'ExtensionDefinition',
          id: NETLIFY_EXTENSION_DEFINITION_ID
        }
      }
    };

    it('returns a list of apps', async () => {
      const loader = {
        getByIds: jest.fn(() => {
          return Promise.resolve({ [NETLIFY_EXTENSION_DEFINITION_ID]: netlifyDefinition });
        })
      };

      const spaceEndpoint = jest.fn(() => {
        return Promise.resolve({ items: [netlifyExtension] });
      });

      const repo = createAppsRepo(loader, spaceEndpoint);
      const result = await repo.getApps();

      expect(loader.getByIds).toBeCalledTimes(1);
      expect(loader.getByIds).toBeCalledWith(
        expect.arrayContaining([NETLIFY_EXTENSION_DEFINITION_ID])
      );

      expect(spaceEndpoint).toBeCalledTimes(1);
      expect(spaceEndpoint).toBeCalledWith({
        method: 'GET',
        path: ['extensions'],
        query: {
          'extensionDefinition.sys.id[in]': expect.stringContaining(NETLIFY_EXTENSION_DEFINITION_ID)
        }
      });

      expect(result).toEqual(
        expect.arrayContaining([
          {
            sys: { type: 'App', id: NETLIFY_APP_ID },
            extensionDefinition: netlifyDefinition,
            extension: netlifyExtension
          }
        ])
      );
    });

    it('does not include the app if definition does not exist', async () => {
      const loader = {
        getByIds: jest.fn(() => Promise.resolve({}))
      };

      const spaceEndpoint = jest.fn(() => {
        return Promise.resolve({ items: [netlifyExtension] });
      });

      const repo = createAppsRepo(loader, spaceEndpoint);
      const result = await repo.getApps();

      expect(loader.getByIds).toBeCalledTimes(1);
      expect(loader.getByIds).toBeCalledWith(
        expect.arrayContaining([NETLIFY_EXTENSION_DEFINITION_ID])
      );

      expect(result).not.toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            sys: { type: 'App', id: NETLIFY_APP_ID }
          })
        ])
      );
    });

    it('includes the app if definition exists but definition does not', async () => {
      const loader = {
        getByIds: jest.fn(() =>
          Promise.resolve({
            [NETLIFY_EXTENSION_DEFINITION_ID]: netlifyDefinition
          })
        )
      };

      const spaceEndpoint = jest.fn(() => {
        return Promise.resolve({ items: [] });
      });

      const repo = createAppsRepo(loader, spaceEndpoint);
      const result = await repo.getApps();

      expect(loader.getByIds).toBeCalledTimes(1);
      expect(loader.getByIds).toBeCalledWith(
        expect.arrayContaining([NETLIFY_EXTENSION_DEFINITION_ID])
      );

      expect(result).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            sys: { type: 'App', id: NETLIFY_APP_ID }
          })
        ])
      );
    });

    it('does not include app extension if there are two extensions for its definition', async () => {
      const loader = {
        getByIds: jest.fn(() =>
          Promise.resolve({
            [NETLIFY_EXTENSION_DEFINITION_ID]: netlifyDefinition
          })
        )
      };

      const spaceEndpoint = jest.fn(() => {
        return Promise.resolve({
          items: [
            { ...netlifyExtension, sys: { type: 'Extension', id: 'e1' } },
            { ...netlifyExtension, sys: { type: 'Extension', id: 'e2' } }
          ]
        });
      });

      const repo = createAppsRepo(loader, spaceEndpoint);
      const result = await repo.getApps();

      expect(loader.getByIds).toBeCalledTimes(1);
      expect(loader.getByIds).toBeCalledWith(
        expect.arrayContaining([NETLIFY_EXTENSION_DEFINITION_ID])
      );

      expect(result).toEqual(
        expect.arrayContaining([
          {
            sys: { type: 'App', id: NETLIFY_APP_ID },
            extensionDefinition: netlifyDefinition
          }
        ])
      );
    });
  });

  describe('getDevApps', () => {
    it('returns fake dev apps for all extension definitions in the current org', async () => {
      const definitions = [
        { sys: { type: 'ExtensionDefinition', id: 'org-def-1' } },
        { sys: { type: 'ExtensionDefinition', id: 'org-def-2' } }
      ];

      const extension = {
        sys: { type: 'Extension', id: 'ext-for-def-2' },
        extensionDefinition: {
          sys: {
            type: 'Link',
            linkType: 'ExtensionDefinition',
            id: 'org-def-2'
          }
        }
      };

      const loader = {
        getAllForCurrentOrganization: jest.fn(() => Promise.resolve(definitions))
      };

      const spaceEndpoint = jest.fn(() => Promise.resolve({ items: [extension] }));

      const repo = createAppsRepo(loader, spaceEndpoint);
      const result = await repo.getDevApps();

      expect(loader.getAllForCurrentOrganization).toBeCalledTimes(1);
      expect(spaceEndpoint).toBeCalledTimes(1);
      expect(spaceEndpoint).toBeCalledWith({
        method: 'GET',
        path: ['extensions'],
        query: {
          'extensionDefinition.sys.id[in]': 'org-def-1,org-def-2'
        }
      });

      expect(result).toEqual([
        {
          sys: { type: 'DevApp', id: 'dev-app_org-def-1' },
          extensionDefinition: definitions[0]
        },
        {
          sys: { type: 'DevApp', id: 'dev-app_org-def-2' },
          extensionDefinition: definitions[1],
          extension
        }
      ]);
    });
  });

  describe('getExtensionDefinitionForApp', () => {
    it('fetches definition by app ID', async () => {
      const loader = { getById: jest.fn(() => 'DEFINITION') };

      const repo = createAppsRepo(loader);
      const result = await repo.getExtensionDefinitionForApp(NETLIFY_APP_ID);

      expect(loader.getById).toBeCalledTimes(1);
      expect(loader.getById).toBeCalledWith(NETLIFY_EXTENSION_DEFINITION_ID);
      expect(result).toBe('DEFINITION');
    });

    it('rethrows loader errors', async () => {
      const loader = {
        getById: jest.fn(() => {
          throw new Error('boom');
        })
      };

      const repo = createAppsRepo(loader);
      expect.assertions(2);
      try {
        await repo.getExtensionDefinitionForApp('netlify');
      } catch (err) {
        expect(loader.getById).toBeCalledTimes(1);
        expect(err.message).toBe('boom');
      }
    });

    it('derives extension definition ID from a dev app ID', async () => {
      const loader = { getById: jest.fn(() => 'DEFINITION') };

      const repo = createAppsRepo(loader);
      const result = await repo.getExtensionDefinitionForApp('dev-app_my-definition');

      expect(loader.getById).toBeCalledTimes(1);
      expect(loader.getById).toBeCalledWith('my-definition');
      expect(result).toBe('DEFINITION');
    });
  });

  describe('getExtensionForExtensionDefinition', () => {
    const definition = {
      sys: {
        type: 'ExtensionDefinition',
        id: 'my-def'
      }
    };

    const makeExtension = extensionId => {
      return {
        sys: { type: 'Extension', id: extensionId || 'my-ext' },
        extensionDefinition: {
          sys: {
            type: 'Link',
            linkType: definition.sys.type,
            id: definition.sys.id
          }
        }
      };
    };

    it('returns extension it there is exactly one extension for definition provided', async () => {
      const spaceEndpoint = jest.fn(() => {
        return Promise.resolve({ items: [makeExtension()] });
      });

      const repo = createAppsRepo(jest.fn(), spaceEndpoint);
      const extension = await repo.getExtensionForExtensionDefinition(definition.sys.id);

      expect(extension.sys.id).toBe('my-ext');
      expect(spaceEndpoint).toBeCalledTimes(1);
      expect(spaceEndpoint).toBeCalledWith({
        method: 'GET',
        path: ['extensions'],
        query: {
          'extensionDefinition.sys.id[in]': 'my-def'
        }
      });
    });

    it('fails if there are no extensions for a definition', async () => {
      const spaceEndpoint = jest.fn(() => {
        return Promise.resolve({ items: [] });
      });

      const repo = createAppsRepo(jest.fn(), spaceEndpoint);
      expect.assertions(2);
      try {
        await repo.getExtensionForExtensionDefinition(definition.sys.id);
      } catch (err) {
        expect(err.message).toMatch(/exactly one Extension/);
        expect(err.extensionCount).toBe(0);
      }
    });

    it('fails if there are multiple extensions for a definition', async () => {
      const spaceEndpoint = jest.fn(() => {
        return Promise.resolve({ items: [makeExtension('e1'), makeExtension('e2')] });
      });

      const repo = createAppsRepo(jest.fn(), spaceEndpoint);
      expect.assertions(2);
      try {
        await repo.getExtensionForExtensionDefinition(definition.sys.id);
      } catch (err) {
        expect(err.message).toMatch(/exactly one Extension/);
        expect(err.extensionCount).toBe(2);
      }
    });
  });
});
