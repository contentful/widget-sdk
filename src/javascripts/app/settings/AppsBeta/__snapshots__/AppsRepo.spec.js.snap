// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppsRepo getAppWidgets should return an object of apps when the endpoint returns good data 1`] = `
Array [
  Object {
    "extensionDefinitionId": "1VchawWvbIClHuMIyxwR5m",
    "icon": "//images.ctfassets.net/lpjm8d10rkpy/2dYQ3x6VMNZJgDXYmMzLBx/b338ad4b50cd0ff31e7157671b801b4d/logomark.svg",
    "id": "netlify",
  },
  Object {
    "extensionDefinitionId": "i43PggcHag2PCrGjYTiIX",
    "icon": "//images.ctfassets.net/lpjm8d10rkpy/4X7O4Q0pIgQZNcONoQrQlp/9262ad9a935fa92e9cacd9207ae0a401/optimizely-logo.svg",
    "id": "optimizely",
  },
  Object {
    "extensionDefinitionId": "zjcnWgBknf9zB7IM9HZjE",
    "icon": "//images.ctfassets.net/lpjm8d10rkpy/3ZH89y9VHHd53v4OQ3BG0V/cc62043a88cd5fb80719687f299e7351/cloudinary_icon_for_white_bg.svg",
    "id": "cloudinary",
  },
  Object {
    "extensionDefinitionId": "5KySdUzG7OWuCE2V3fgtIa",
    "icon": "//images.ctfassets.net/lpjm8d10rkpy/3vNJZRjthatucy2mBJMy1u/0e5d501207427780021e7580cc384761/bynder.5e9e457b9d.svg",
    "id": "bynder",
  },
  Object {
    "extensionDefinitionId": "342Q2DqCjmsdN5BJCEPkrJ",
    "icon": "//images.ctfassets.net/lpjm8d10rkpy/7v1wryVbq6dooI1OqbkIpg/32e947099d5253c56fa344fc7aadfbff/app-logo.svg",
    "id": "image-focal-point",
  },
]
`;

exports[`AppsRepo getMarketplaceApps should return an object of apps when the endpoint returns good data 1`] = `
Array [
  Object {
    "actionList": Array [
      Object {
        "info": "Build widget will be removed from sidebars of all content types",
        "negative": true,
      },
      Object {
        "info": "This will not modify your Netlify sites. You may want to remove unneeded deploy notifications and build hooks from your Netlify site.",
        "negative": false,
      },
    ],
    "author": Object {
      "icon": "//images.ctfassets.net/lpjm8d10rkpy/4DxiiBjixHZVjc69WpJX95/4708b0bdc8e713faf69a667f8266d190/472182",
      "name": "Contentful",
      "url": "https://www.contentful.com",
    },
    "categories": Array [
      "Featured",
      "Website Hosting",
    ],
    "description": "### About the app

The Netlify App gives content editors control over static site builds and makes the process more transparent.

---

### This app provides the following functionality:

- triggers Netlify builds with a simple click in the sidebar
- displays build status so editors don't preview outdated sites
- brings Netlify functionality into the Contentful web app
- makes it possible to configure on a per-content-type level

---

### To use this app, you need:

- a static site based on a service such as Gatsby, Middleman or Hugo
- your site configured to load content from Contentful and its code pushed to a remote repository, such as GitHub or Gitlab
- a Netlify setup where the site is connected to the remote repository and enabled for continuous deployment

---

### Screenshots

![Netlify config](//images.ctfassets.net/lpjm8d10rkpy/4Rf8yoGPw4AQwdMboiVwhF/a0a9a8b8e73a5ace1fdf78adebc24a49/apps-netlify-site-setup.b4a24a69c0.png)

![Netlify sidebar](//images.ctfassets.net/lpjm8d10rkpy/RrSqXTcdlpfkPt7sJzkEb/57172d0fd3eaa36a5ab90e7d183ed171/apps-netlify-sidebar.0ffd73e12b.png) ",
    "extensionDefinition": Object {
      "locations": Array [
        "app",
        "entry-sidebar",
      ],
      "name": "Netlify",
      "src": "http://localhost:1234",
      "sys": Object {
        "id": "1VchawWvbIClHuMIyxwR5m",
        "type": "ExtensionDefinition",
      },
    },
    "extensionDefinitionId": "1VchawWvbIClHuMIyxwR5m",
    "flagId": "basic_apps",
    "icon": "//images.ctfassets.net/lpjm8d10rkpy/2dYQ3x6VMNZJgDXYmMzLBx/b338ad4b50cd0ff31e7157671b801b4d/logomark.svg",
    "id": "netlify",
    "installed": true,
    "links": Array [
      Object {
        "title": "Netlify Docs",
        "url": "https://www.netlify.com/docs/",
      },
      Object {
        "title": "Documentation for the Netlify app",
        "url": "https://www.contentful.com/developers/docs/extensibility/apps/netlify/",
      },
    ],
    "permissions": "__Netlify app__ is asking for permissions to access this Contentful environment.

The app will be able to manage this environment and any content in it.

You can revoke access at any time by uninstalling the app.",
    "permissionsExplanation": undefined,
    "tagLine": "Easily build and preview Netlify sites directly from Contentful ",
    "title": "Netlify",
  },
]
`;
